[ML] Add lazy assignment job config option

droberts195 opened the pull request:
This change adds:

- A new option, allow_lazy_open, to anomaly detection jobs
- A new option, allow_lazy_start, to data frame analytics jobs

Both work in the same way: they allow a job to be
opened/started even if no ML node exists that can
accommodate the job immediately. In this situation
the job waits in the opening/starting state until ML
node capacity is available. (The starting state for data
frame analytics jobs is new in this change.)

Additionally, the ML nightly maintenance tasks now
creates audit warnings for ML jobs that are unassigned.
This means that jobs that cannot be assigned to an ML
node for a very long time will show a yellow warning
triangle in the UI.

A final change is that it is now possible to close a job
that is not assigned to a node without using force.
This is because previously jobs that were open but
not assigned to a node were an aberration, whereas
after this change they'll be relatively common.

droberts195 committed:
[ML] Add lazy assignment job config option

This change adds:

- A new option, allow_lazy_open, to anomaly detection jobs
- A new option, allow_lazy_start, to data frame analytics jobs

Both work in the same way: they allow a job to be
opened/started even if no ML node exists that can
accommodate the job immediately. In this situation
the job waits in the opening/started state until ML
node capacity is available.

Additionally, the ML nightly maintenance tasks now
creates audit warnings for ML jobs that are unassigned.
This means that jobs that cannot be assigned to an ML
node for a very long time will show a yellow warning
triangle in the UI.

droberts195 committed:
Add daily auditing for unassigned jobs

droberts195 committed:
Merge branch 'master' into allow_lazy_job_assignment

elasticmachine commented:
Pinging @elastic/ml-core (:ml)

droberts195 committed:
Add a unit test

droberts195 committed:
Allow graceful close of unassigned jobs

This change is similar to the one that was made
to allow stopping unassigned datafeeds in #39034
because if unassigned jobs are now going to be
the norm rather than the exception and remain
unassigned for long periods then it's very
frustrating to not be allowed to easily close
them.

droberts195 committed:
Changing analytics assignment audit message to be more consistent with others

And updating tests to match

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlDailyMaintenanceService.java:
```diff
@@ -113,17 +122,34 @@ private synchronized void scheduleNext() {
+    private void auditUnassignedMlTasks(ClusterState state) {
+        PersistentTasksCustomMetaData tasks = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
+        if (tasks != null) {
+            mlAssignmentNotifier.auditMlTasks(state.nodes(), tasks, tasks, true);
```
It's a bit odd that we're passing the same `tasks` as both the `previousTasks` and the `CurrentTasks` here. Assuming this works properly, should we encapsulate that in `MlAssignmentNotifier` by making the current `auditMlTasks` method `private` and having another `public` one that only accepts a single `tasks` arg?

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:
```diff
@@ -116,30 +116,39 @@ protected void doExecute(Task task, CloseJobAction.Request request, ActionListen
+                                                // The listener here can be a no-op, as waitForJobClosed() already waits for
+                                                // these persistent tasks to disappear.
+                                                persistentTasksService.sendRemoveRequest(jobTask.getId(),
+                                                    ActionListener.wrap(r -> {}, e -> {}));
```
Should we not handle an error in case the removal of the persistent task fails?

droberts195 replied:
I'll add logging of the error at this point in the code.  To do anything more would mean completely refactoring the way the class waits for the per-job requests.

The overall request will time out instead of reporting success if removing the persistent task for an unassigned job fails.  This is similar to what happens if removing the persistent task for an assigned job fails after we've done all the other steps.  Effectively the change is making treatment of assigned and unassigned jobs more similar.

dimitris-athanasiou commented on x-pack/plugin/src/test/resources/rest-api-spec/test/ml/jobs_crud.yml:
```diff
@@ -113,8 +113,9 @@
 ---
-"Test put job with model_memory_limit as string":
-
+"Test put job with model_memory_limit as string and lazy open":
```
It would be great to add a similar test for df-analytics

dimitris-athanasiou reviewed (commented):
Leaving a few comments but it looks great!

droberts195 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/DataFrameAnalyticsConfig.java:
```diff
@@ -56,6 +56,7 @@ public static Builder builder() {
     private static final ParseField MODEL_MEMORY_LIMIT = new ParseField("model_memory_limit");
     private static final ParseField CREATE_TIME = new ParseField("create_time");
     private static final ParseField VERSION = new ParseField("version");
+    private static final ParseField ALLOW_LAZY_START = new ParseField("allow_lazy_start");
```
For the equivalent setting in anomaly detection jobs, it can be added to the modules the SIEM team is using to configure their ML jobs.  Then every time those jobs are started they'll pick up the setting.  Basically with the work flows that are envisaged for this new setting we think that it will be something that's always desired or not desired for a particular job, and not desired for some invocations but not others.

Then for the data frame analytics setting specifically I made it symmetrical to the way things are done for anomaly detection jobs.

droberts195 committed:
Merge branch 'master' into allow_lazy_job_assignment

droberts195 committed:
Some review comments

droberts195 committed:
Add a STARTING start for DF analytics

Trying to add an integration test for lazy assignment
showed how unfriendly it is to report starting but
unassigned analytics tasks as STOPPED.  We will cause
enormous customer confusion without a STARTING state.

droberts195 committed:
Add an integration test for lazy assignment of df analytics

droberts195 committed:
Fix docs

droberts195 committed:
Merge branch 'master' into allow_lazy_job_assignment

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java:
```diff
@@ -433,6 +433,7 @@ public boolean test(PersistentTasksCustomMetaData.PersistentTask<?> persistentTa
                 case STOPPING:
                     exception = ExceptionsHelper.conflictStatusException("the task has been stopped while waiting to be started");
                     return true;
+                case STARTING:
```
I think this means that when a user calls start with `lazy` mode they would not get a response until the task is actually assigned, right? We probably don't want that.

droberts195 replied:
In the case where the job is too big to be assigned this check higher up the method will return early:

```
            // This means we are awaiting a new node to be spun up, ok to return back to the user to await node creation
            if (assignment != null && assignment.equals(JobNodeSelector.AWAITING_LAZY_ASSIGNMENT)) {
                return true;
            }
```

The only time the code will get to the `STARTING` case in the `switch` is during the tiny fraction of a second when a job has been successfully assigned but the call to update its task state has not yet completed.  Previously in this situation the state was reported as `STOPPED`, so I thought it best to put `STARTING` alongside `STOPPED` in the `switch` to maintain previous behaviour.

(I think the integration tests confirm this works as I described, but just to double check I tested this manually with a job with a `3tb` `model_memory_limit` and `allow_lazy_start: true` on my laptop and the start endpoint returned success instantly.)

It's equivalent to this bit of `TransportOpenJobAction`: https://github.com/elastic/elasticsearch/blob/f52f2aef9707bdf93d9a785452429137d2689b5a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportOpenJobAction.java#L561-L563

But since it's clearly confusing I'll add some comments.

dimitris-athanasiou replied:
Right, got it.

droberts195 committed:
Add comments

dimitris-athanasiou reviewed (approved):
LGTM

droberts195 committed:
Merge branch 'master' into allow_lazy_job_assignment
